src/repository.ts

```diff
@@ -0,0 +1,107 @@
+import * as Storage from "./components/storage/implementation"
+import * as TypeChecks from "./common/type-checks.js"
+
+
+export type RepositoryOptions = {
+  storage: Storage.Implementation
+  storageName: string
+}
+
+
+export default abstract class Repository<T> {
+
+  dictionary: Record<string, T>
+  memoryCache: T[]
+  storage: Storage.Implementation
+  storageName: string
+
+
+  constructor({ storage, storageName }: RepositoryOptions) {
+    this.memoryCache = []
+    this.dictionary = {}
+    this.storage = storage
+    this.storageName = storageName
+  }
+
+  static async create<T>(options: RepositoryOptions) {
+    // @ts-ignore
+    const repo = new this.prototype.constructor(options)
+
+    repo.memoryCache = await repo.getAll()
+    repo.dictionary = repo.toDictionary(repo.memoryCache)
+
+    return repo
+  }
+
+  async add(itemOrItems: T | T[]): Promise<void> {
+    const items = Array.isArray(itemOrItems) ? itemOrItems : [ itemOrItems ]
+
+    this.memoryCache = [ ...this.memoryCache, ...items ]
+    this.dictionary = this.toDictionary(this.memoryCache)
+
+    await this.storage.setItem(
+      this.storageName,
+      // TODO: JSON.stringify(this.memoryCache.map(this.toJSON))
+      this.memoryCache.map(this.toJSON).join("|||")
+    )
+  }
+
+  clear(): Promise<void> {
+    this.memoryCache = []
+    this.dictionary = {}
+
+    return this.storage.removeItem(this.storageName)
+  }
+
+  find(predicate: (value: T, index: number) => boolean): T | null {
+    return this.memoryCache.find(predicate) || null
+  }
+
+  getByIndex(idx: number): T | null {
+    return this.memoryCache[ idx ]
+  }
+
+  async getAll(): Promise<T[]> {
+    const storage = await this.storage.getItem(this.storageName)
+    const storedItems = TypeChecks.isString(storage)
+      // TODO: ? - Need partial JSON decoding for this
+      ? storage.split("|||").map(this.fromJSON)
```

Comment:
I'd like something better here, but don't know how to approach this.
I thought I could `JSON.stringify` the whole collection, but I need to call `fromJSON` on every item 🤔